src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -0,0 +1,254 @@
+/*-------------------------------------------------------------------------
+ *
+ * insert_select_planner.c
+ *
+ * Planning logic for INSERT..SELECT.
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "distributed/citus_ruleutils.h"
+#include "distributed/errormessage.h"
+#include "distributed/insert_select_planner.h"
+#include "distributed/multi_executor.h"
+#include "distributed/multi_logical_planner.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/multi_router_planner.h"
+#include "distributed/pg_dist_partition.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/parsenodes.h"
+#include "optimizer/planner.h"
+#include "parser/parsetree.h"
+#include "parser/parse_coerce.h"
+#include "parser/parse_relation.h"
+#include "utils/lsyscache.h"
+
+
+static DeferredErrorMessage * DeferErrorIfCoordinatorInsertSelectUnsupported(
+	Query *insertSelectQuery);
+static Query * WrapSubquery(Query *subquery);
+
+
+/*
+ * CreatteCoordinatorInsertSelectPlan creates a query plan for a SELECT into a
+ * distributed table. The query plan can also be executed on a worker in MX.
+ */
+MultiPlan *
+CreateCoordinatorInsertSelectPlan(Query *parse)
+{
+	Query *insertSelectQuery = copyObject(parse);
+	Query *selectQuery = NULL;
+
+	RangeTblEntry *selectRte = ExtractSelectRangeTableEntry(insertSelectQuery);
+	RangeTblEntry *insertRte = ExtractInsertRangeTableEntry(insertSelectQuery);
+	Oid targetRelationId = insertRte->relid;
+
+	ListCell *selectTargetCell = NULL;
+	ListCell *insertTargetCell = NULL;
+
+	MultiPlan *multiPlan = CitusMakeNode(MultiPlan);
+	multiPlan->operation = CMD_INSERT;
+
+	multiPlan->planningError =
+		DeferErrorIfCoordinatorInsertSelectUnsupported(insertSelectQuery);
+
+	if (multiPlan->planningError != NULL)
+	{
+		return multiPlan;
+	}
+
+	selectQuery = selectRte->subquery;
+
+	/*
+	 * Wrap the SELECT as a subquery if the INSERT...SELECT has CTEs or the SELECT
+	 * has top-level set operations.
+	 *
+	 * We could simply wrap all queries, but that might create a subquery that is
```

Comment:
Hopefully one day I'll update `ReorderInsertSelectTargetList()` to handle top level union 😄